src/v/raft/buffered_protocol.cc

```diff
@@ -0,0 +1,394 @@
+            const auto request_latency = now - sent_ts;
+            _last_reply_timestamp = now;
+            _hist.record(request_latency);
+            reply_f.forward_to(std::move(reply_promise));
+        });
+}
+
+bool append_entries_queue::can_buffer_next_request(size_t size) const {
+    return _requests.empty()
+           || (_buffered_bytes + size) < _max_buffered_bytes();
+}
+bool append_entries_queue::is_idle() const {
+    static constexpr auto queue_idle_timeout = 30s;
+    return _requests.empty() && inflight_requests() == 0
+           && _last_sent_timestamp < clock_type::now() - queue_idle_timeout;
+}
+
+ss::future<result<append_entries_reply>> append_entries_queue::append_entries(
+  append_entries_request r, rpc::client_opts opts) {
+    // or wait until we can buffer next request, this will propagate back
+    // pressure to caller
+    auto sz = r.total_size();
+    return _dispatched.wait([this, sz] { return can_buffer_next_request(sz); })
+      .then([this, r = std::move(r), opts = std::move(opts)]() mutable {
+          /// consensus is no longer responsible for tracking memory usage and
+          /// dispatch ordering after this point
+          opts.resource_units.reset();
+          _buffered_bytes += r.total_size();
+          _requests.emplace_back(std::move(r), std::move(opts));
+
+          _new_requests.signal();
+          return _requests.back().reply.get_future();
+      });
+}
+ss::future<> append_entries_queue::stop() {
+    vlog(_logger.debug, "stopping append entries queue");
+    _new_requests.broken();
+    _dispatched.broken();
+    _inflight_requests_sem.broken();
+    for (auto& r : _requests) {
+        r.reply.set_value(raft::errc::shutting_down);
+    }
+    _requests.clear();
+    _public_metrics.clear();
+    _internal_metrics.clear();
+    return _gate.close().finally(
+      [this] { vlog(_logger.debug, "stopped queue"); });
+}
+
+void append_entries_queue::log_status() const {
+    vlog(
+      _logger.info,
+      "inflight requests: {}, buffered requests: {}, buffered_bytes: {}",
+      inflight_requests(),
+      _requests.size(),
+      _buffered_bytes);
+}
+
+void append_entries_queue::setup_internal_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+    sm::label_instance target_node_id_label("target_node_id", _target_node);
+    _internal_metrics.add_group(
+      prometheus_sanitize::metrics_name("raft::buffered::protocol"),
+      {sm::make_gauge(
+         "inflight_requests",
+         [this] { return inflight_requests(); },
+         sm::description("Number of append entries requests that were sent to "
+                         "the target node and are awaiting responses."),
+         {target_node_id_label}),
+       sm::make_gauge(
+         "buffered_bytes",
+         [this] { return _buffered_bytes; },
+         sm::description("Total size of append entries requests in the queue"),
```

Comment:
These kinds of metrics are good as they give you kind of an "average" view into the system. 

However at higher scrape frequencies like in cloud (60s) you miss a lot of micro behaviour that happens in between the scrapes. 

For that I think we'd want a counter how often requests fail to acquire the units immediately so you can always know whether any requests got blocked on this in the metrics interval.